Refactor strings to use an inference manager object

ajreynol opened the pull request:
This organizes the theory of strings so that the processing of inferences is handled by a central, customized inference manager object, which manages whether inferences are facts/lemmas/conflicts. It can be seem as a wrapper around the output channel of TheoryStrings and the equality engine of TheoryStrings.

This required cleaning up a few other minor things (mkAnd, utilities related to proxy variables) to enable this split.

The behavior of the code is mostly unchanged.

This is work towards #1881.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+  Trace("strings-lemma") << "Strings::Lemma " << c << " SPLIT : " << lemma_or
+                         << std::endl;
+  d_pending_lem.push_back(lemma_or);
+  sendPhaseRequirement(eq, preq);
```
The splits statistic is missing. Can we maybe add it to `OutputChannelStrings`?

ajreynol replied:
This stat was never really used and is not very important, hence I'd like to remove it for simplicity.

4tXJ7f replied:
In that case, we should probably get rid of the variable completely.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+    return d_ee.getRepresentative(t);
+  }
+  return t;
+}
```
IMHO, we should only have an implementation here or in `TheoryStrings` and not in both.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+  return t;
+}
+
+bool OutputChannelStrings::hasTerm(Node a) { return d_ee.hasTerm(a); }
```
Ditto.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+
+bool OutputChannelStrings::hasTerm(Node a) { return d_ee.hasTerm(a); }
+
+bool OutputChannelStrings::areEqual(Node a, Node b)
```
Ditto.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+  return false;
+}
+
+bool OutputChannelStrings::areDisequal(Node a, Node b)
```
Ditto.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+
+void OutputChannelStrings::doPendingLemmas()
+{
+  if (!hasConflict())
```
In the original code, we don't do that if we don't have pending lemmas. Is this change intentional? It seems like it could make a difference if we have pending phase requirements but no pending lemmas.

ajreynol replied:
Yes, the previous code was a bit nonsensical.  Lemmas and phase requirements essentially should be processed in parallel (phase requirements typically say what polarity to decide for literals that appear in lemmas).  The new code makes it so that we don't do anything if we are in conflict.

4tXJ7f replied:
That's fair, thanks for the explanation.

4tXJ7f commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+#include "theory/uf/equality_engine.h"
+
+#include <climits>
+#include <deque>
```
Those includes should come before the project includes.

4tXJ7f commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+#define CVC4__THEORY__STRINGS__OUTPUT_CHANNEL_H
+
+#include "context/cdhashset.h"
+#include "theory/uf/equality_engine.h"
```
Add missing includes for `Context`, `Node`, `std::vector`, `OutputChannel`, `std::map`.

4tXJ7f commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+   *
+   * This requests that ( a = b V a != b ) is sent on the output channel as a
+   * lemma. We additionally request that a phase requirement the equality a=b
+   * to polarity preq.
```
The last sentence does not seem to parse for me.

4tXJ7f commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+   * of) literals that each are explainable, i.e. they already hold in the
+   * equality engine of this class.
+   */
+  void sendInfer(Node eq_exp, Node eq, const char* c);
```
Why are the two methods above `protected` instead of `private`?

4tXJ7f commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+  /** The list of pending literals to assert to the equality engine */
+  std::vector<Node> d_pending;
+  /** A map from the literals in the above vector to their explanation */
+  std::map<Node, Node> d_pending_exp;
```
Since this is a new class, can we rename the variables to be camelCase to follow the style guide?

4tXJ7f commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+   * unproc = {}.
+   * In particular, since says that the information content of n essentially
+   * says that x = v2, while the first and third conjunctions can be dropped
+   * since they effectively give the definition of the proxy variables.
```
The sentence above does not seem to parse for me.

4tXJ7f commented on src/theory/strings/output_channel.cpp:
```diff
@@ -0,0 +1,448 @@
+  d_pending_req_phase.clear();
+}
+
+bool OutputChannelStrings::hasConflict() const { return d_parent.d_conflict; }
```
Wouldn't it be more natural to move `d_parent.d_conflict` to `OutputChannelStrings`?

ajreynol replied:
There would be a circular include.

4tXJ7f replied:
Hm, I am not quite following. `theory_strings.h` already includes `inference_manager.h` https://github.com/CVC4/CVC4/pull/3076/files#diff-452ff9fd1723599f909e96f23d6f366dR27, no?

ajreynol replied:
Yes, so inference_manager cannot include theory_strings.h. Hence, this function needs to be in the cpp of inference_manager.

4tXJ7f replied:
I think there was a bit of a misunderstanding here :) I was suggesting to move the `d_conflict` variable to `InferenceManager` ;) Not too important, we can potentially address this in a later PR.

ajreynol replied:
Ah.  Yes I considered this.

I think in the end I'm going to split another class off of TheoryStrings, which relates to the "state" of the solver, which includes things like read queries to equality engine and the normal form information.  Otherwise I think the inference manager will become too big of a class.  The intention of this class is to mostly act as a means of performing the "output" of strategies and not means of tracking the state of the solver.

I am following a very similar approach in refactoring sets, see InferenceManager, SetsState (https://github.com/CVC4/CVC4/compare/master...ajreynol:setsCardSplit).

ajreynol commented on src/theory/strings/theory_strings_utils.cpp:
```diff
@@ -0,0 +1,50 @@
+namespace CVC4 {
+namespace theory {
+namespace strings {
+namespace utils {
```
I was following https://github.com/CVC4/CVC4/blob/master/src/theory/bv/theory_bv_utils.h, although I am not strongly for/against this namespace.

4tXJ7f replied:
I would prefer that there is a `utils` folder insides `src/theory/{bv,strings}` in that case but I agree that that discussion is out-of-scope for this PR. We should discuss this though.

ajreynol commented on src/theory/strings/theory_strings_utils.h:
```diff
@@ -0,0 +1,43 @@
+ * Make the conjunction of nodes in a. Removes duplicate conjuncts, returns
+ * true if a is empty, and a single literal if a has size 1.
+ */
+Node mkAnd(std::vector<Node>& a);
```
I think this is a bigger discussion, that we've debated a lot in the past and won't be easy to quickly resolve. I'd like to revisit this but its out of scope for the PR.

This PR is regardless progress towards this since it is making strings/BV more uniform as they both will now have a utils file.

4tXJ7f replied:
I tend to agree with this and yes, we should discuss this again. To have a utils file is at least a step forward :)

ajreynol commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+/*********************                                                        */
+/*! \file output_channel.h
```
I've changed the name to inference_manager and the class name to InferenceManager.

4tXJ7f replied:
That sounds good to me.

ajreynol commented on src/theory/strings/output_channel.h:
```diff
@@ -0,0 +1,308 @@
+ * calls will be processed as:
+ *
+ * (1) Internally in the strings solver, via calls to the equality engine. We
+ * refer to these literals as "facts",
```
What I call "facts" does refer to literals. 

A "conclusion" indeed may be an arbitrary formula, which can be processed to multiple literal "facts".

4tXJ7f replied:
Right. In the context, it is still a bit confusing IMHO because the text says: 

> whether the _conclusions_ will be processed [...] We refer to these literals as "facts"

It could be understood that the conclusion has to be a simple literal in case (1), which is AFAIK not true. Maybe we could say:

> We refer to literals in these conclusions as "facts"

ajreynol commented:
I've addressed all points of the review except those that I've commented on. 

I have left the strings utilities in the `strings::utils::` namespace.  My main reason why I like it is that it is a way of marking utility functions in the code. That is, seeing `utils::mkAnd` immediately tells me that its from `strings/theory_strings_utils.h`, whereas if it were a global function in the same namespace it would be a bit more confusing.  I'm also basing this on my experience in the bit-vector theory with theory_bv_utils, where I generally like having the `utils::` prefix there too.

This is up for discussion.

4tXJ7f commented on src/theory/strings/inference_manager.h:
```diff
@@ -0,0 +1,292 @@
+   * vars = { x },
+   * subs = { v2 },
+   * unproc = {}.
+   * In particular, this says that the information content of n essentially
```
```suggestion
   * In particular, this says that the information content of n is essentially
```

aniemetz commented:
> I've addressed all points of the review except those that I've commented on.
> 
> I have left the strings utilities in the `strings::utils::` namespace. My main reason why I like it is that it is a way of marking utility functions in the code. That is, seeing `utils::mkAnd` immediately tells me that its from `strings/theory_strings_utils.h`, whereas if it were a global function in the same namespace it would be a bit more confusing. I'm also basing this on my experience in the bit-vector theory with theory_bv_utils, where I generally like having the `utils::` prefix there too.
> 
> This is up for discussion.

I agree. I prefer having util functions in the utils:: namespace. I also prefer to always use `utils::` explicitly in the code (no `using utils::...`).
